common/diagnostics/Diagnostics.java

```diff
@@ -18,63 +18,146 @@
-                                               ErrorReporter errorReporter) {
-        if (diagnostics != null) {
-            LOG.debug("Skipping re-initialising diagnostics");
-            return;
+    public static class Noop extends Diagnostics {
+        private Noop() {
+            super(null, null, null);
+        }
+
+        public static synchronized void initialise() {
+            Sentry.init(options -> options.setEnabled(false));
+            diagnostics = new Diagnostics.Noop();
         }
-        Sentry.init(options -> {
-            options.setEnabled(enable);
-            options.setDsn(diagnosticsURI);
-            options.setEnableTracing(true);
-            options.setSendDefaultPii(false);
-            options.setRelease(releaseName(distributionName, version));
-        });
-        User user = new User();
-        user.setUsername(serverID);
-        Sentry.setUser(user);
-        diagnostics = new Diagnostics(errorReporter);
+
+        @Override
+        public void mayStartServing(@Nullable SslContext sslContext, ChannelInboundHandlerAdapter... middleware) {}
+        @Override
+        public void submitError(Throwable error) {}
+        @Override
+        public void requestAttempt(Metrics.NetworkRequests.Kind kind) {}
+        @Override
+        public void requestSuccess(Metrics.NetworkRequests.Kind kind) {}
+        @Override
+        public void setCurrentCount(Metrics.CurrentCounts.Kind kind, long value) {}
     }
 
-    public static synchronized void initialiseNoop() {
-        Sentry.init(options -> options.setEnabled(false));
-        diagnostics = new Diagnostics(new ErrorReporter.NoopReporter());
+    public static class Core extends Diagnostics {
+        protected Core(Metrics metrics, StatisticReporter statisticReporter, MonitoringServer monitoringServer) {
+            super(metrics, statisticReporter, monitoringServer);
+        }
+
+        public static synchronized void initialise(
+                String serverID, String distributionName, String version,
+                boolean errorReportingEnable, String errorReportingURI,
+                boolean statisticsReportingEnable, String statisticsReportingURI,
+                boolean monitoringEnable, int monitoringPort
+        ) {
+            if (diagnostics != null) {
+                LOG.debug("Skipping re-initialising diagnostics");
+                return;
+            }
+
+            initSentry(serverID, distributionName, version, errorReportingEnable, errorReportingURI);
+
+            Metrics metrics = new Metrics(serverID, distributionName, version);
+            StatisticReporter statisticReporter = initStatisticReporter(statisticsReportingEnable, statisticsReportingURI, metrics);
+            MonitoringServer monitoringServer = initMonitoringServer(monitoringEnable, monitoringPort, metrics);
+
+            diagnostics = new Core(metrics, statisticReporter, monitoringServer);
```

Comment:
So these are just `null`able? Where do we do the null check to do nothing if the reporter/server is null?